plugins/Toolbox/src/PackagesModel.py

```diff
@@ -0,0 +1,101 @@
+        self.addRoleName(Qt.UserRole + 3, "name")
+        self.addRoleName(Qt.UserRole + 4, "version")
+        self.addRoleName(Qt.UserRole + 5, "author_name")
+        self.addRoleName(Qt.UserRole + 6, "author_email")
+        self.addRoleName(Qt.UserRole + 7, "description")
+        self.addRoleName(Qt.UserRole + 8, "icon_url")
+        self.addRoleName(Qt.UserRole + 9, "image_urls")
+        self.addRoleName(Qt.UserRole + 10, "download_url")
+        self.addRoleName(Qt.UserRole + 11, "last_updated")
+        self.addRoleName(Qt.UserRole + 12, "is_bundled")
+
+        # List of filters for queries. The result is the union of the each list of results.
+        self._filter = {}  # type: Dict[str, str]
+
+    def setMetadata(self, data):
+        self._metadata = data
+        self._update()
+
+    def _update(self):
+        items = []
+
+        for package in self._metadata:
+            items.append({
+                "id": package["package_id"],
+                "type": package["package_type"],
+                "name": package["display_name"],
+                "version": package["package_version"],
+                "author_name": package["author"]["name"],
+                "author_email": package["author"]["email"] if "email" in package["author"] else None,
+                "description": package["description"],
+                "icon_url": package["icon_url"] if "icon_url" in package else None,
+                "image_urls": package["image_urls"] if "image_urls" in package else None,
+                "download_url": package["download_url"] if "download_url" in package else None,
+                "last_updated": package["last_updated"] if "last_updated" in package else None,
+                "is_bundled": package["is_bundled"] if "is_bundled" in package else False
+            })
+
+        # Filter on all the key-word arguments.
+        for key, value in self._filter.items():
+            if "*" in value:
+                key_filter = lambda candidate, key = key, value = value: self._matchRegExp(candidate, key, value)
+            else:
+                key_filter = lambda candidate, key = key, value = value: self._matchString(candidate, key, value)
+            items = filter(key_filter, items)
+
+        # Execute all filters.
+        filtered_items = list(items)
+
+        filtered_items.sort(key = lambda k: k["name"])
+        self.setItems(filtered_items)
+
+    ##  Set the filter of this model based on a string.
+    #   \param filter_dict \type{Dict} Dictionary to do the filtering by.
+    def setFilter(self, filter_dict: Dict[str, str]) -> None:
+        if filter_dict != self._filter:
+            self._filter = filter_dict
+            self._update()
+
+    @pyqtProperty("QVariantMap", fset = setFilter, constant = True)
+    def filter(self) -> Dict[str, str]:
+        return self._filter
+
+    # Check to see if a container matches with a regular expression
+    def _matchRegExp(self, metadata, property_name, value):
```

Comment:
Maybe we should consider moving this function and `_matchString` to `ListModel` by this point...

Reply:
I'm going to leave it since the filtering may actually be removed eventually once more specific API calls are possible.